webknossos/webknossos/geometry/nd_bounding_box.py

```diff
@@ -0,0 +1,507 @@
+            ) from err
+
+    def get_3d(self, attr_name: str) -> Vec3Int:
+        axes = ("x", "y", "z")
+        attr_3d = []
+
+        for axis in axes:
+            index = self.axes.index(axis)
+            attr_3d.append(getattr(self, attr_name)[index])
+
+        return Vec3Int(attr_3d)
+
+    def set_3d(self, attr_name: str, value: Vec3IntLike) -> VecInt:
+        value = Vec3Int(value)
+        axes = ("x", "y", "z")
+        modified_attr = getattr(self, attr_name).to_list()
+
+        for i, axis in enumerate(axes):
+            index = self.axes.index(axis)
+            modified_attr[index] = value[i]
+
+        return VecInt(modified_attr)
+
+    def _check_compatibility(self, other) -> None:
+        if self.axes != other.axes:
+            raise ValueError(
+                f"Operation with two bboxes is only possible if they have the same axes. {self.axes} != {other.axes}"
+            )
+
+    def padded_with_margins(
+        self, margins_left: VecIntLike, margins_right: Optional[VecIntLike] = None
+    ) -> None:
+        raise NotImplementedError()
+
+    def intersected_with(
+        self, other: "NDBoundingBox", dont_assert: bool = False
+    ) -> "NDBoundingBox":
+        """If dont_assert is set to False, this method may return empty bounding boxes (size == (0, 0, 0))"""
+
+        self._check_compatibility(other)
+        topleft = self.topleft.pairmax(other.topleft)
+        bottomright = self.bottomright.pairmin(other.bottomright)
+        size = (bottomright - topleft).pairmax(VecInt.zeros(len(self.size)))
+
+        intersection = attr.evolve(self, topleft=topleft, size=size)
+
+        if not dont_assert:
+            assert (
+                not intersection.is_empty()
+            ), f"No intersection between bounding boxes {self} and {other}."
+
+        return intersection
+
+    def extended_by(self, other: "NDBoundingBox") -> "NDBoundingBox":
+        self._check_compatibility(other)
+        if self.is_empty():
+            return other
+        if other.is_empty():
+            return self
+
+        topleft = self.topleft.pairmin(other.topleft)
+        bottomright = self.bottomright.pairmax(other.bottomright)
+        size = bottomright - topleft
+
+        return attr.evolve(self, topleft=topleft, size=size)
+
+    def is_empty(self) -> bool:
+        return not self.size.is_positive(strictly_positive=True)
+
+    def in_mag(self, mag: Mag) -> "NDBoundingBox":
+        mag_vec = mag.to_vec3_int()
+
+        assert (
+            self.get_3d("topleft") % mag_vec == Vec3Int.zeros()
+        ), f"topleft {self.topleft} is not aligned with the mag {mag}. Use BoundingBox.align_with_mag()."
+        assert (
+            self.get_3d("bottomright") % mag_vec == Vec3Int.zeros()
+        ), f"bottomright {self.bottomright} is not aligned with the mag {mag}. Use BoundingBox.align_with_mag()."
+
+        new_topleft = self.set_3d("topleft", Vec3Int(self.get_3d("topleft") // mag_vec))
```

Comment:
Yes NDBoundingBox is immutable. I also think, it would be great to have better names for the `set_3d()` and `get_3d()` method. The `set_3d` method returns a tuple of int of a requested parameter, where the x,y and z value are inserted in the desired position.

So if I have `bbox = NDBoundingBox(topleft=(0,0,0,10,20), size=(7,5,100,100,100), axes=(t,q,x,z,y), index=(1,2,3,4,5))`
`bbox.get_3d("topleft")` returns `(0, 20, 10)`
and
`bbox.set_3d("topleft", (0, 10, 20))` returns `(0, 0, 0, 20, 10)`